\documentclass{article}
\usepackage[utf8]{inputenc}
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage[table,xcdraw]{xcolor}
\usepackage{hyperref}
\usepackage{svg}
\usepackage{geometry}
\geometry{a4paper, margin=1.1in}



\date{}
\begin{document}
\title{TEPIC 2 - An extended framework for transcription factor binding prediction and integrative epigenomic analysis}
\maketitle

TEPIC is a versatile framework for the analysis of transcription factor (TF) binding and offers several machine learning approaches for integrative analysis of predicted transcription factor binding sites (TFBS) and gene-expression data.
Briefly, TEPIC offers:
\begin{itemize}
\item Annotation of user defined regions with TF affinities using TRAP and a variety of provided TF-motifs,
\item Aggregation of TF affinities to TF-gene scores,
\item Computation of statistical scores such as peak-length, peak-count or peak-signal per gene,
\item Inclusion of long range chromatin contacts,
\item Discretization of continuous TF affinities using a background distribution into a binary measure for TF-binding,
\item Linear regression analysis to infer key transcriptional regulators within one sample,
\item Logistic regression classifier to suggest key transcriptional regulators between samples,
\item Generate input for DREM to infer important TFs from temporal epigenomic and gene expression data.
\end{itemize}

These points are illustrated in Figure \ref{FigOverview}.
This document provides a brief introduction into the functionality of TEPIC and the machine learning approaches. 
\begin{figure}[h]
\includegraphics[width=\textwidth]{workflow_TEPIC_General.pdf}
\caption{Overview on the workflows supported by TEPIC. Input to the submodules is shown in blue while the output is shown in black.}
\label{FigOverview}
\end{figure}
\newpage
\section{Introduction to TEPIC}
\subsection{Motivation}
TF are essential players in transcriptional regulation. To understand their function, it is essential to know their binding sites genome-wide. 
Although TFBS can be inferred from ChIP-seq experiments, several in-silico approaches have been developed as well to overcome the burden and complexity of wet-lab experiments.
Especially computational methods considering epigenetics data in the prediction have been used succesfully to predict TFBS. 
The main advantage of considering epigenetics data for the task of TF binding prediction is that the number of false positive predictions can be reduced \cite{pmid21106904}.
One way of incorporating epigenetics data is to reduce the genomic search space to a few candidate regions of TF binding. 
As shown before, genome-wide candidate sites for TF binding can be determined by open-chromatin experiments \cite{pmid25294828,pmid25086003,pmid22072382,pmid23424114}, e.g. peaks or footprints in DNase1-seq data, 
and/or by considering Histone marks \cite{pmid25489339,pmid25086003}, e.g. H3K4me3. 

Here, we compute TF affinities for currated sets of \textit{Position Specific Energy Matrices (PSEMs)} using \textit{TRAP} \cite{pmid17098775} which is based on a biophysical model of TF binding \cite{von1986specificity}. 
A major advantage of affinity based predictions compared to hit-based methods like Fimo \cite{Grant16022011} is that 
low-affinity binding sites can be included \cite{pmid27899623,pmid17098775}. Using the \textit{TEPIC} method, we compute TF gene scores by aggregating TF predictions calculated for a user defined set of candidate regions.
The scores, either per peak/region or gene, can be interpreted as a quantitative measurement of TF binding. 

\subsection{Collection of TF-motifs}
We obtained \textit{Position Count Matrices (PCMs)} from JASPAR \cite{pmid26531826}, which is also including data from Uniprobe \cite{pmid25378322}, HOCOMOCO \cite{pmid23175603} and the Kellis Lab ENCODE Motif database \cite{pmid24335146}.

There are three folder containing Position specific energy matrices (PSEMs): Our current collection of PSEMs \textit{PWMs/2.1}. The previously used motifs are provided in the folders \textit{PWMs/2.0} and \textit{PWMs/1.0}. 
TF motifs used in the original TEPIC manuscript are stored in the file \\\textit{PWMs/1.0/pwm\_vertebrates\_jaspar\_uniprobe\_original.PSEM}.

In detail, the current collection contains from the \textit{JASPAR 2018 Core} database:
\begin{itemize}
\item 579 PSEMs for vertebrates
\item 176 PSEMs for fungi
\item 26 PSEMs for nematodes
\item 489 PSEMs for plants
\item 1 PSEM for urochordates
\item 133 PSEMs for insects
\end{itemize}

Additionally, we provide species specific collections of JASPAR matricies:
\begin{itemize}
\item 3 PSEMs for Antirrhinum majus
\item 5 PSEMs for Arabidopsis lyrata
\item 440 PSEMs for Arabidopsis thaliana
\item 22 PSEMs for Caenorhabditis elegans
\item 132 PSEMs for Drosophila melanogaster
\item 1 PSEMs for Fragaria x ananassa
\item 7 PSEMs for Gallus gallus
\item 6 PSEMs for Glycine max
\item 1 PSEM for Halocynthia roretzi
\item 459 PSEMs for Homo sapiens
\item 1 PSEM for Hordeum vulgare
\item 1 PSEM for Medicago truncatula
\item 1 PSEM for Meleagris gallopavo
\item 157 PSEMs for Mus musculus
\item 1 PSEM for Neurospora crassa
\item 1 PSEM for Nicotiana
\item 4 PSEMs for Orcytolagus
\item 7 PSEMs for Oryza sativa
\item 1 PSEM for Petunia x hybrida
\item 1 PSEM for Phaeodactylum tricornutum
\item 9 PSEMs for Physcomitrella patens
\item 3 PSEMs for Pisum sativum
\item 1 PSEM for Populus trichocarpa
\item 2 PSEMs for Rattus norvegicus
\item 2 PSEMs for Rattus rattus
\item 176 PSEMs for Saccaromyces cerevisiae
\item 2 PSEMs for Solanum lycopersicum
\item 1 PSEM for Triticum aestivum
\item 4 PSEMs for Xenopus laevis
\item 8 PSEMs for Zea mays
\end{itemize}

All JASPAR matrices can be found in \textit{PWMs/2.1/JASPAR\_PSEMs}

From HOCOMOCO we provide $402$ motifs for homo sapiens and $358$ for mus musculus, available in \textit{PWMs/2.1/HOCOMOCO\_PSEMs}

The Kellis set contains $58$ motifs, stored in \textit{PWMs/2.1/Kellis\_PSEMs}.

Additionally we provide non-redundant collections for homo sapiens and mus musculus considering motifs from all three sources:
\begin{itemize}
\item 561 PSEMS for homo sapiens
\item 380 PSEMs for mus musculus
\end{itemize}

The matrices are stored in the folder \textit{PWMs/2.1/Merged\_PSEMs}

Furthermore, we used a motif clustering approach (7), to merge similar motifs of the files containing matrices from all three sources. This lead to
\begin{itemize}
\item 483 PSEMs for homo sapiens
\item 306 PSEMs for mus musculus
\end{itemize}
The matrices are stored in the folder \textit{PWMs/2.1/Clustered\_PSEMs}

Files holding the length of the PSEMs are provided too.


\subsection{Converting position count matrices to position specific energy matrices}
As mentioned above, \textit{TRAP} computes TF affinities that are based on a biophysical model of TF binding.
Therefore \textit{PCMs} have to be converted to \textit{Position Specific Energy Matrices (PSEMs)} such that they can be used in \textit{TRAP}.
Intuitively, \textit{PSEMs} represent the mismatch energy of a given motif. For a detailed explanation and motivation of the energy based score, please check \cite{pmid17098775}.
A \textit{PCM M} is converted to a \textit{PSEM E} according to:
\begin{align}
E_{i,j}=\frac{1}{\lambda}log(\frac{M_{max,j}}{M_{i,j}}b_{i,j}), \\
M_{max,j}=\max\limits_{i\in\{A,C,G,T\}}(M_{i,j}).
\end{align}
The parameter $\lambda$ is used for scaling the mismatch energies and $b_{i,j}$ denotes the background frequency of the nucleotide $i$ with respect to the most frequent nucleotide at position $j$. 
This conversion formula is part of the mismatch energy postulated in formula (4) in \cite{pmid17098775}. 
By definition, if $j=max$, than $E_{i,j}=0$, as there should be no mismatch energy for the best possible sequence match. 
Note that, during conversion, a pseudo count $pc = 1$ is added to each $M_{i,j}$. 

The conversion is done by a C++ tool provided by the authors of \textit{TRAP}. This is also included in the \textit{TEPIC} repository.
As suggested in \cite{pmid17098775}, we use the following parameters for the conversion:
\begin{itemize}
\item $\lambda=0.7$
\item $m=0.584$
\item $n=-5.66$
\end{itemize}
The parameters \textit{slope m} and \textit{intercept n} are used to compute a matrix specific parameter $R_0$ that combines the concentration of the corresponding TF and the
equilibrium constant of the binding reaction with its optimal binding site as defined in \cite{pmid17098775}. The authors of \textit{TRAP} found a linear approximation for $R_0$ with:
\begin{align}
ln(R_0)=m*|M|+n,
\end{align}
where $|M|$ denotes the length of the \textit{PCM} as above.
\bigskip
\\Further, we exploit species specific GC-content values:
\begin{itemize}
\item \textit{homo sapiens }$=0.41$
\item \textit{mus musculus }$=0.42$
\item \textit{rattus norvegicus }$=0.42$
\item \textit{drosophila melanogaster }$=0.43$
\item \textit{caenorhabditis elegans }$=0.36$
\end{itemize}
In all other cases, a default GC-content of $0.42$ is used.

\subsection{Computing TF gene scores}
Using our collections of \textit{PSEMs}, \textit{TRAP} computes TF binding affinities in all user provided regions 
that could be found in the reference genomes of the respective species and 
overlap with a window of user defined size $w$ that is centered at the most $5'$ TSS of all annotated genes in the considered organism. 
Then, TF-gene scores are computed by incorporating all candidate binding sites within the window centered around the $5'$ TSS of genes in the final score. 
The contribution of the individual sites is weighted by their distance to the selected TSS with an exponential decay function \cite{pmid19995984}.
Formally, the TF gene score $a_{g,i}$ for gene $g$ and TF $i$ is computed as
\begin{align}
a_{g,i}^{w}&=\sum_{p \in P_{g,w}} a_{p,i}e^{-\frac{d_{p,g}}{d_0}},
\end{align}
where $a_{p,i}$ is the affinity of TF $i$ in peak $p$, the set $P_{g,x}$ contains all open-chromatin peaks
in a window of size $w$ around gene $g$, $d_{p,g}$ is the distance from the center of peak $p$ to the TSS of gene $g$, and $d_0$ is a constant fixed at $5000$bp \cite{pmid19995984}.
Additionally, affinities can be normalized by peak(and motif)-length during the computation of gene-TF scores:
\begin{align}
a_{g,i}^{w}&=\sum_{p \in P_{g,w}} {\frac{a_{p,i}}{|p|-|m|}e^{-\frac{d_{p,g}}{d_0}}},
\end{align}
where $|p|$ is the length of peak $p$, $|m_i|$ is
the length of the motif of TF $i$, with an extra count of $1$.
If the signal within a peak should be directly considered in the gene-TF score, we compute:
\begin{align}
a_{g,i}^{w}&=\sum_{p \in P_{g,w}} {\frac{a_{p,i}}{|p|-|m|}s_{p}e^{-\frac{d_{p,g}}{d_0}}},
\end{align}
where $s_p$ is the per base signal in peak $p$. This computation can be done with and without length normalisation of the affinities. 
The workflow of TEPIC is depicted in Figure \ref{workflowFig}.

In addition to the TF gene scores, TEPIC can compute features for peak length ($pl_g$), peak count ($pc_g$), and peak signal($ps_g$) following the same scoring formulation as for
TF affinities:
\begin{align}
pl_g&=\sum_{p \in P_{g,w}}|p|e^{-\frac{d_{p,g}}{d_0}}, \\
pc_g&=\sum_{p \in P_{g,w}}e^{-\frac{d_{p,g}}{d_0}}, \\
ps_g&=\sum_{p \in P_{g,w}}s_{p}e^{-\frac{d_{p,g}}{d_0}},
\end{align}
where $|p|$ is the length of $p$. These features can be used for example to assess the influence of chromatin accessibility on gene expression without considering TF binding predictions. 

Furthermore, TEPIC can compute a TF-specific affinity cut-off derived from either user-defined, or randomly generated sequences, to distinguish likely bound sites from unbound sites. These scores
can be used to come-up with a binary TF-gene assignment. Further details on this mode are provided in Section \ref{EPIC-DREM}.
\begin{figure}[h!]
\begin{center}
\includegraphics[width=\textwidth]{Workflow.png}
\end{center}
\caption{The general workflow of \textit{TEPIC} is as follows: 
Data of an open-chromatin or Histone modification ChIP-seq experiment needs to be preprocessed to generate a genome segmentation, 
either by peak for footprint calling. 
Using the segmentation, \textit{TEPIC} applies \textit{TRAP} in all regions of interest, and computes TF gene scores using exponential decay to reweigh 
TF binding predictions in open-chromatin regions based on their distance to a genes TSS.} 
\label{workflowFig}
\end{figure}

With version $2.2$ of TEPIC, we introduced support for the inclusion of long range chromatin conformation capture data. In addition to the promoter centric windows used before, we
calculate TF affinities $a_{g,i}*$ and peak scores $pl_g*, pc_g*, ps_g*$ for all DHSs residing  in genomic loci looping into the promoter region of a gene, summarizied in $P_{g,V_g}$, where $V_g$ is the set of all regions looped into the promoter region of gene $g$:
\begin{align}
a_{g,i}&=\sum_{p \in P_{g,V_g}} a_{p,i},\\
pl_g*&=\sum_{p \in P_{g,V_g}}|p|, \\
pc_g*&=\sum_{p \in P_{g,V_g}}, \\
ps_g*&=\sum_{p \in P_{g,V_g}}s_{p}.
\end{align}
Note that scores computed for $p \in P_{g,V_g}g$ are never considering the exponential decay as a direct interaction of the respective sites with the promoter region of gene $g$ has been determined by chromatin conformation capture experiments.

\subsection{Required input}
To compute TF gene scores, a user needs to specify:
\begin{itemize}
\item a reference genome (-g option),
\item a set of \textit{PSEMs} (-p option),
\item a set of genomic regions in BED format (-b option).
\item a gtf file containing the genome annotation (-a option).
\end{itemize}
Note that the chromosome identifiers in the BED file must match the identifiers used in the reference genomes. Otherwise they can not be considered. 
Special care should be taken for \textit{caenorhabditis elegans}, as Roman digits are used for enumeration of chromosomes.

\subsection{Output}
TEPIC outputs:
\begin{enumerate}
\item TF affinities for all selected \textit{PSEMs} in the regions provided by the user that passed the filtering step (\textit{\*\_Affinity.txt}). 
\item (Length normalized) TF gene scores for all selected \textit{PSEMs} calculated as described above (optionally including peak features) (\textit{\*\_Affinity\_Gene\_View.txt}). 
\item A meta data file listing all used parameters (\textit{amd.tsv}).
\item Optionally a seperate file containing the signal information in peaks (\textit{\*\_Peak\_Coverage.txt}). 
\item TF affinities with all values below an inferred threshold set to zero (\textit{\*\_Thresholded\_Affinity.txt})
\item A sparse representation linking TF to genes (\textit{\*\_Sparse\_Affinity\_Gene\_View.txt})
\end{enumerate}

\newpage
\section{Identification of key transcriptional regulators using epigenetics data (INVOKE)}
Epigenetics data contains a wealth of information on gene regulation. It was shown that especially
data on open-chromatin is well suited to build predictive models of gene-expression \cite{pmid27899623,pmid22955983,pmid25231769,pmid22954627}.
Interpreting these models allows the inference of regulators that may play a key role in gene-expression regulation.
\bigskip
\\Hear, we offer an integrated analysis of epigenetics data, e.g. open-chromatin data (DNase1-seq, ATAC-seq, NOMe-seq) and gene-expression data
to suggest key transcriptional regulators in the analysed sample.
\bigskip
\\Note that, although incorporating epigenetic data greatly improved the performance of TF binding predictions, both computing TF binding predictions and linking TFs to genes are still unsolved problems and all predictions
should be seen as suggestions and not as the absolute truth.
\bigskip
\\The \textit{INVOKE} analysis is split up into two main steps. 
\begin{enumerate}
\item Computing TF gene scores on the basis of epigenetic data using \textit{TEPIC} (see above).
\item Learning a linear regression model to predict gene expression from TF gene scores computed in (1).
\end{enumerate}

\subsection{Linear regression to predict gene expression}
\subsubsection{Motivation}
In order to learn about potentially important regulators, we build a linear, interpretable regression model, 
comparable to methods proposed in \cite{pmid27899623,pmid22955983,pmid25231769,pmid22954627}.
Here, we use TF gene scores computed with \textit{TEPIC} as features in a linear regression setup to predict gene expression.
In such a \textit{per sample} approach, we stick to the simplifying assumption that all genes are regulated similarly. 
Features with a high regression coefficient can be suggested to be key regulators in the analysed sample, as they seem to affect the expression
of a large portion of the genes under consideration. However, the results of this method should be seen as suggestions for possible regulators and not as the absolute truth. 

Details on the learning setup and on the available regularization methods are provided in the next section.

\subsubsection{Available regularization methods}
We offer three different regularization techniques:
\begin{itemize}
\item Lasso:
\begin{align}
 \hat{\beta}&=\underset{\beta}{\arg\,min} ||y-X\beta||^2 + ||\beta||,
\end{align}
\item Ridge:
\begin{align}
 \hat{\beta}&=\underset{\beta}{\arg\,min} ||y-X\beta||^2 + ||\beta||^2,
\end{align}
\item Elastic net:
\begin{align}
 \hat{\beta}&=\underset{\beta}{\arg\,min} ||y-X\beta||^2 + \alpha||\beta||^2 + (1-\alpha)||\beta||,
\end{align}
\end{itemize}
where, $\beta$ represents the regression coefficient vector, $\hat{\beta}$ represents the estimated coefficients, $X$ is the feature matrix, $y$ is the response vector, and
the parameter $\alpha$ controls the distribution between Ridge and Lasso penalty in the elastic net.
\bigskip
\\Using Lasso regularization, models are sparse and can be learned very fast. 
But, Lasso cannot properly deal with correlated features, e.g. instead of distributing the coefficients among them, only one is selected. Also, Lasso solutions are not stable and therefore should be interpreted with caution. Nevertheless, Lasso regularization is good to get a first impression of model performance. 

The disadvantage of Ridge regression is that it cannot produce sparse models (many coefficients being exactly 0), which may hinder interpretability. 

Elastic net regularization was designed to overcome the limitations of both regularization techniques mentioned above.
It resolves the correlation between features by distributing the feature weights among them, and simultaneously leads to sparse and stable models \cite{Zou05regularizationand}. 
However, learning a model using elastic net penalty is slower than using either only Lasso or Ridge regularization.

\subsubsection{Details on the learning setup}
The data matrix $X$, containing TF gene scores, and the response vector $y$, containing gene expression values, are log-transformed, 
with a pseudo-count of $1$, centered and scaled to fit them as. 
Regression coefficients are computed in an inner cross validation,
the $\alpha$ parameter of elastic net regularization is optimized with a default step size of $0.1$.

We offer two ways to use our learning pipeline:
\begin{enumerate}
\item{Learn a model for feature interpretation without computing performance measures:
In order to provide a time efficient way of obtaining an interpretable model and to prevent a potential loss of information
by considering only a portion of the full data set for model training, the regression coefficients are determined on the entire data set.}
\item{Learn a model for feature interpretation and compute model performance: 
Nested cross-validation is used to learn the models and to assess their performance. 
Per default, $20\%$ of the data are used as test data and $80\%$ are used as training data. 
Model performance is assessed in an outer cross validation. 
We report the mean pearson correlation, the mean spearman correlation, and the mean squared error over the outer folds as measures of model performance.
Additionally, a model is learned on the entire data set as described in (1) for interpretation of the coefficients.}
\end{enumerate}
All parameters mentioned in this section can be changed by the user. The learning process is sketched in Figure \ref{learningFig}.

\subsubsection{Required input}
In addition to the input required for the computation of TF gene scores in TEPIC, a file containing gene expression data must be provided.
This file should be structured such that column $1$ contains the gene identifiers and column $2$ holds expression values.

\subsubsection{Output and hints for interpretation}
The user is always provided with the following files:
\begin{itemize}
\item a list of regression coefficients computed on the entire data set,
\item a bar plot showing the regression coefficients with an absolute value $> 0.025$.
\end{itemize}
The larger a regression coefficient, the stronger is the inferred effect of the corresponding TF on gene expression. Positive coefficients suggest an activating influence of TFs,
negative coefficients suggest an inhibiting effect. 

If model performance was assessed, the following is available in addition:
\begin{itemize}
\item a summary on model performance containing the aforementioned measures (pearson correlation, spearman correlation, mean squared error),
\item a list of regression coefficients determined in the outer cross validation,
\item a heatmap visualizing the regression coefficients determined in the outer cross validation for at most the top $10$ positive and negative features, sorted according to their median.
\item an image showing a box plot for pearson and spearman correlation respectively.
\item scatter plots showing the predicted vs the measured gene expression for each outer cross validation fold.
\end{itemize}
The heatmap can be easily used to judge model performance, as it shows the regression coefficients of all outer-cross validation runs. The box plots provide further insights into model performance and stability across the outer folds of the cross validation.
\vspace{1cm}
\begin{figure}[h!]
\begin{center}
\includegraphics[width=\textwidth]{Learning.pdf}
\end{center}
\caption{Overview of the learning process. 
The left part of the Figure describes the assessment of model performance in a $y$-fold outer cross validation and $6$-fold inner cross validation. 
The right hand site illustrates model training on the entire data set, again using a $6$-fold inner cross validation for parameter learning.}
\label{learningFig}
\end{figure}

\newpage
\section{Differential analysis to identify novel transcriptional regulators for differentially expressed genes (DYNAMITE)}
Although a variety of methods have been proposed to generate genome-wide TF binding predictions [\cite{pmid27899623,pmid22072382,pmid23424114,pmid25086003}] and to establish \textit{TF to tissue} 
associations [\cite{pmid22955983,pmid19995984,pmid27899623}], systematic, feasible, and easy to use ways of linking TFs to distinct genes are rare. 

In addition to the \textit{INVOKE} analysis, we propose a method to infer the most likely transcriptional regulators for a set of differentially expressed genes. 
We use TF scores, computed using \textit{TEPIC}, and logistic regression to identify TFs that have explanatory power to distinguish between up- and down-regulated genes. 

\subsection{Input}
To run \textit{DYNAMITE}, a user most provide candidate regions of TF binding for two groups of samples, $A$ and $B$, e.g. control and disease. 
These can be derived, for example, by open chromatin experiments such as DNase-seq. 
It is essential that the candidate regions reflect the characteristics of chromatin organization in the analysed tissues. 
In addition, a list of differential expressed genes between two groups as well as log2 fold changes of the expression are needed. 

\section*{Method}
Our method consists of two parts: (1) gene-TF score computation, and (2) identification of key TFs. 
\subsection*{Step 1: Computing Gene-TF Scores}
Using TEPIC, we compute gene-TF scores $g_{ij}$ for all differentially expressed genes $i$ and distinct TFs $j$ considering the provided candidate regions for all replicates $a$ of group $A$ and for all replicates $b$ of group $B$. 
As a result, gene-TF matrices $M_k$ for all replicates of both groups are obtained. To account for biological variation among the replicates, we compute two matrices $M_A$, $M_B$ holding the mean gene-TF scores among all replicates of a group, where
\begin{align}
 M_{A_{ij}}= \frac{\sum_{a \in A}{M_{a_{ij}}}}{|A|},
 \\M_{B_{ij}}= \frac{\sum_{b \in B}{M_{b_{ij}}}}{|B|}.
\end{align}
Using matrices $M_A$ and $M_B$ we compute a matrix $R_{AB}$ that holds the ratios of gene-TF scores for all genes and all TFs:
\begin{equation}
    R_{AB_{ij}}=\frac{M_{A_{ij}}}{M_{B_{ij}}}.
\end{equation}
Thus, $R_{AB}$ represents the changes in TF binding between groups $A$ and $B$ on a gene level.
The feature computation is sketched in Figure \ref{TF-Gene-Score_Computation}.
\begin{figure}[h!]
\centering
\includegraphics[width=0.75\textwidth]{TF_Affinities.png}
\caption{Computation of differential TF features between two groups.}
\label{TF-Gene-Score_Computation}
\end{figure}
\subsection{Step 2: Identification of Key Transcription Factors}
To identify those TFs that can explain the differential expression state of as many genes as possible, we build a logistic regression classifier. 
We use matrix $R_{AB}$ computed in Step $1$ as the feature matrix $X$, and a binary vector of gene expression changes as response $y$. 
An example is shown in Figure \ref{Log-Reg-Example}.
\begin{figure}[h!]
\centering
\includegraphics[width=0.5\textwidth]{Example_Matrix.png}
\caption{Example for a matrix used as input to the logistic regression. The column \textit{Expression Changes} is used as response, while the affinity ratios \textit{TFx} are used as features.}
\label{Log-Reg-Example}
\end{figure}
We perform logistic regression with elastic net regularisation (11).
As above, we tune the parameter $\alpha$ that distributes the weight between lasso and ridge penalty in a grid search with user defined step-size between $0$ and $1$.

Model parameters are learned in an inner cross validation, while the accuracy of our classifier can be assessed through an outer cross validation. This is the
same learning paradigm that is described for the \textit{INVOKE} analysis (Figure \ref{learningFig}). We use the entire dataset for model training and to interpret the regression coefficients.
TFs that correspond to features with a non-zero regression coefficient can be seen as being essential to explain the observed expression differences and should be further investigated.

\section{Output}
Model performance is reported in a \textit{txt} file and visually in a bar plot using mean test and training accuracy as well as the F1 measure.
A heatmap shows the regression coefficients in the outer cross validation folds. 
Additionally, we report confusion matrices for the outer cross validation folds.
We generate a bar-plot with the regression coefficients of all TFs selected in the final model.
A positive coefficient is used by the model to predict genes as upregulated, a negative coefficient is related to genes that are predicted as downregulated. The interpretation
of the model can be simplified if the user makes sure that both TF ratios and gene expression fold changes are computed in the same order. 

We provide an additional script to generate further plots per feature that can help to understand the model. As shown in Figure \ref{ExampleAnalysisOfDynamite},
density plots, and scatter plots are generated to help elucidating why a particular feature was selected by the model.
\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{Feature_overview_TEMvsTN_Peak_Counts.pdf}
\caption{Example for an automatically created feature analysis Figure generated on the example data provided in the repository. The density plots show the distribution of TF affinities, the
scatter plot relates the TF affinities to the observed expression changes. The miniature heatmap shows the regression coefficients determined during the outer cross validation.}
\label{ExampleAnalysisOfDynamite}
\end{figure}

\newpage
\section{Determine important transcriptional regulators from time serires data (EPIC-DREM)}
\label{EPIC-DREM}
\textit{EPIC-DREM} is a combination of \textit{TEPIC}  and the \textit{Dynamic Regulatory Events Miner (DREM)} \cite{pmid17224918}.
Instead of using static ChIP-seq data, which is provided in \textit{DREM 2.0}, we suggest to use time-point specific TF binding predictions
based on time-dependent epigenomic profiles. Thereby, \textit{DREM} can infer regulators that can be linked to expression changes at distinct points in time.
We have shown that using the predicted, dynamic TF binding events is superior to the static data included in \textit{DREM}.

In order generate a sparse input matrix for \textit{DREM}, we devised a strategy to threshold TF affinities based on a set of background sequences.
These can be either chosen automatically or be provided by the user. Please check the README file for detailed options. 

In Figure \ref{epicdrem}, we illustrate how TF affinities can be discretized and illustrate their usage in \textit{DREM}. Note that \textit{DREM}
is not included in the \textit{TEPIC} repository. It is available \href{http://www.sb.cs.cmu.edu/drem/}{online}.
\begin{figure}[h!]
\begin{center}
\includegraphics[width=\textwidth]{epicdrem.pdf}
\end{center}
\caption{Overview on the EPIC-DREM approach: Note that in this Figure, different time points are indicated by different colors. First, epigenetic data, e.g. DNase1 experiments, are conducted for different points (a).
Next, putative TF binding sites are identified by peak and/or footprint calling (b,c) and annotated with TF affinities (d). 
From the putative binding sites, a random set of genomic regions is chosen (e) and annotated with TF affinities as well (f). By applying a p-value cut-off on the distribution of TF affinities calculated on the random regions (g),
a suitable, TF specific affinity threshold is chosen to discretize the original TF affinities (h). Using the default TEPIC TF-gene score formulation (i), a TF-gene interaction matrix (j) is computed. Together with gene expression data (k),
the sparse matrix (j) can be used as input for DREM (l) to identify potential key regulators of expression changes in time series data.}
\label{epicdrem}
\end{figure}

\subsection{Thresholded TF affinities}
In some applications it is required to make a binary decision whether a factor is binding or not. 
To infer this information from TF affinities, \textit{TEPIC} allows the computation of a TF specific affinity threshold by calculating TF affinities on a randomly selected set of genomic regions. When selected by TEPIC, these regions show similar characteristics compared to the provided regions (GC content and length). 
Alternatively, a set of background regions can be provided by the user.
By applying a user defined p-value on the distribution of affinities computed on the random regions, a threshold is chosen. 
Per TF, all affinities that are smaller than the selected threshold, are set to zero, thus a sparse matrix with TF-gene interactions can be generated. 

\subsubsection{Required input}
In addition to the input mentioned above, a reference genome in 2bit format is required. 
Optionally, the user can provide a bed file containing background regions. 
These replace the automated generation of background sequences. 

\subsubsection{Output}
The following output files are generated in addition:
\begin{enumerate}
\item TF affinities for all selected \textit{PSEMs} in the regions provided by the user that passed the filtering step, where all affinities below the TF specific thresholds are set to 0.
\item (Length normalized) TF gene scores for all selected \textit{PSEMs} calculated as described above (optionally including peak features) using the thresholded affinities.
\item A sparse representation of TF gene interactions.
\end{enumerate}

Either (2) or (3) can be combined with RNA-seq data and used as input for \textit{DREM}.
\newpage
\bibliographystyle{plain}
\bibliography{Description}
\end{document} 
